parlai/utils/fp16.py

```diff
@@ -85,68 +87,186 @@ def has_overflow(grad_norm):
     return False
 
 
-###################################################
-## APEX Wrapper
-###################################################
+class SafeFP16Optimizer(torch.optim.Optimizer):
```

Comment:
what exactly is a safe fp16 optimizer?

Reply:
same as the apex fp16 optimizer, just ours